File: LR1/Tree.cpp
#include "Tree.h"
#include <iostream>
#include <fstream>
#include <queue>
#include <stack>

using namespace std;

void ClearTree(TNode* node) {
    if (!node) return;
    ClearTree(node->left);
    ClearTree(node->right);
    delete node;
}

void TADD(Tree& tree, const string& value, bool silent) {
    TNode* newNode = new TNode(value, nullptr, nullptr, nullptr);
    
    if (!tree.root) {
        tree.root = newNode;
    } else {
        queue<TNode*> q;
        q.push(tree.root);
        TNode* parent = nullptr;
        
        while (!q.empty()) {
            parent = q.front();
            q.pop();
            
            if (!parent->left) {
                parent->left = newNode;
                newNode->parent = parent;
                break;
            } else if (!parent->right) {
                parent->right = newNode;
                newNode->parent = parent;
                break;
            } else {
                q.push(parent->left);
                q.push(parent->right);
            }
        }
    }
    
    tree.size++;
    if (!silent) {
        cout << "Добавлен в дерево: " << value << endl;
    }
}

bool TFIND(const Tree& tree, const string& value) {
    if (!tree.root) {
        cout << "Дерево пусто!\n";
        return false;
    }
    
    queue<TNode*> q;
    q.push(tree.root);
    
    while (!q.empty()) {
        TNode* current = q.front();
        q.pop();
        
        if (current->data == value) {
            cout << "Элемент '" << value << "' найден в дереве" << endl;
            return true;
        }
        
        if (current->left) q.push(current->left);
        if (current->right) q.push(current->right);
    }
    
    cout << "Элемент '" << value << "' не найден в дереве" << endl;
    return false;
}

bool TISFULL(const Tree& tree) {
    if (!tree.root) return true;
    
    queue<TNode*> q;
    q.push(tree.root);
    bool foundNonFullNode = false;
    
    while (!q.empty()) {
        TNode* current = q.front();
        q.pop();
        
        bool hasZeroChildren = (!current->left && !current->right);
        bool hasTwoChildren = (current->left && current->right);
        
        if (!hasZeroChildren && !hasTwoChildren) {
            return false;
        }
        
        if (foundNonFullNode && !hasZeroChildren) {
            return false;
        }
        
        if (hasZeroChildren) {
            foundNonFullNode = true;
        }
        
        if (current->left) q.push(current->left);
        if (current->right) q.push(current->right);
    }
    
    return true;
}

void TLEVELORDER(const Tree& tree) {
    if (!tree.root) {
        cout << "Дерево пусто!\n";
        return;
    }
    
    cout << "Level-order: ";
    queue<TNode*> q;
    q.push(tree.root);
    
    while (!q.empty()) {
        TNode* current = q.front();
        q.pop();
        cout << current->data << " ";
        
        if (current->left) q.push(current->left);
        if (current->right) q.push(current->right);
    }
    cout << endl;
}

void TPRINT_STRUCTURE(const Tree& tree) {
    if (!tree.root) {
        cout << "Дерево пусто!\n";
        return;
    }
    
    cout << "Структура дерева:\n";
    queue<TNode*> q;
    q.push(tree.root);
    int level = 0;
    int nodesInLevel = 1;
    
    while (!q.empty()) {
        int nextLevelCount = 0;
        cout << "Уровень " << level << ": ";
        
        for (int i = 0; i < nodesInLevel; i++) {
            if (q.empty()) break;
            
            TNode* current = q.front();
            q.pop();
            
            if (current) {
                cout << current->data << " ";
                q.push(current->left);
                q.push(current->right);
                nextLevelCount += 2;
            } else {
                cout << "NULL ";
                q.push(nullptr);
                q.push(nullptr);
                nextLevelCount += 2;
            }
        }
        
        cout << endl;
        level++;
        nodesInLevel = nextLevelCount;
        
        bool hasNonNull = false;
        queue<TNode*> temp = q;
        while (!temp.empty()) {
            if (temp.front()) {
                hasNonNull = true;
                break;
            }
            temp.pop();
        }
        
        if (!hasNonNull) break;
    }
}

void TPREORDER(const Tree& tree) {
    if (!tree.root) {
        cout << "Дерево пусто!\n";
        return;
    }
    
    cout << "Pre-order: ";
    stack<TNode*> s;
    s.push(tree.root);
    
    while (!s.empty()) {
        TNode* current = s.top();
        s.pop();
        cout << current->data << " ";
        
        if (current->right) s.push(current->right);
        if (current->left) s.push(current->left);
    }
    cout << endl;
}

void TINORDER(const Tree& tree) {
    if (!tree.root) {
        cout << "Дерево пусто!\n";
        return;
    }
    
    cout << "In-order: ";
    stack<TNode*> s;
    TNode* current = tree.root;
    
    while (current || !s.empty()) {
        while (current) {
            s.push(current);
            current = current->left;
        }
        
        current = s.top();
        s.pop();
        cout << current->data << " ";
        
        current = current->right;
    }
    cout << endl;
}

void TPOSTORDER(const Tree& tree) {
    if (!tree.root) {
        cout << "Дерево пусто!\n";
        return;
    }
    
    cout << "Post-order: ";
    stack<TNode*> s1, s2;
    s1.push(tree.root);
    
    while (!s1.empty()) {
        TNode* current = s1.top();
        s1.pop();
        s2.push(current);
        
        if (current->left) s1.push(current->left);
        if (current->right) s1.push(current->right);
    }
    
    while (!s2.empty()) {
        cout << s2.top()->data << " ";
        s2.pop();
    }
    cout << endl;
}

void TCLEAR(Tree& tree) {
    ClearTree(tree.root);
    tree.root = nullptr;
    tree.size = 0;
    cout << "Дерево очищено" << endl;
}

File: LR1_GO/Tree.go
package main

import (
	"container/list"
	"fmt"
)

func clearTree(node *TNode) {
	if node == nil {
		return
	}
	clearTree(node.Left)
	clearTree(node.Right)
}

func TADD(tree *Tree, value string, silent bool) {
	newNode := &TNode{Data: value, Left: nil, Right: nil, Parent: nil}
	
	if tree.Root == nil {
		tree.Root = newNode
	} else {
		queue := list.New()
		queue.PushBack(tree.Root)
		var parent *TNode
		
		for queue.Len() > 0 {
			front := queue.Front()
			parent = front.Value.(*TNode)
			queue.Remove(front)
			
			if parent.Left == nil {
				parent.Left = newNode
				newNode.Parent = parent
				break
			} else if parent.Right == nil {
				parent.Right = newNode
				newNode.Parent = parent
				break
			} else {
				queue.PushBack(parent.Left)
				queue.PushBack(parent.Right)
			}
		}
	}
	
	tree.Size++
	if !silent {
		fmt.Printf("Добавлен в дерево: %s\n", value)
	}
}

func TFIND(tree *Tree, value string) bool {
	if tree.Root == nil {
		fmt.Println("Дерево пусто!")
		return false
	}
	
	queue := list.New()
	queue.PushBack(tree.Root)
	
	for queue.Len() > 0 {
		front := queue.Front()
		current := front.Value.(*TNode)
		queue.Remove(front)
		
		if current.Data == value {
			fmt.Printf("Элемент '%s' найден в дереве\n", value)
			return true
		}
		
		if current.Left != nil {
			queue.PushBack(current.Left)
		}
		if current.Right != nil {
			queue.PushBack(current.Right)
		}
	}
	
	fmt.Printf("Элемент '%s' не найден в дереве\n", value)
	return false
}

func TISFULL(tree *Tree) bool {
	if tree.Root == nil {
		return true
	}
	
	queue := list.New()
	queue.PushBack(tree.Root)
	foundNonFullNode := false
	
	for queue.Len() > 0 {
		front := queue.Front()
		current := front.Value.(*TNode)
		queue.Remove(front)
		
		hasZeroChildren := (current.Left == nil && current.Right == nil)
		hasTwoChildren := (current.Left != nil && current.Right != nil)
		
		if !hasZeroChildren && !hasTwoChildren {
			return false
		}
		
		if foundNonFullNode && !hasZeroChildren {
			return false
		}
		
		if hasZeroChildren {
			foundNonFullNode = true
		}
		
		if current.Left != nil {
			queue.PushBack(current.Left)
		}
		if current.Right != nil {
			queue.PushBack(current.Right)
		}
	}
	
	return true
}

func TLEVELORDER(tree *Tree) {
	if tree.Root == nil {
		fmt.Println("Дерево пусто!")
		return
	}
	
	fmt.Print("Level-order: ")
	queue := list.New()
	queue.PushBack(tree.Root)
	
	for queue.Len() > 0 {
		front := queue.Front()
		current := front.Value.(*TNode)
		queue.Remove(front)
		fmt.Printf("%s ", current.Data)
		
		if current.Left != nil {
			queue.PushBack(current.Left)
		}
		if current.Right != nil {
			queue.PushBack(current.Right)
		}
	}
	fmt.Println()
}

func TPRINT_STRUCTURE(tree *Tree) {
	if tree.Root == nil {
		fmt.Println("Дерево пусто!")
		return
	}
	
	fmt.Println("Структура дерева:")
	queue := list.New()
	queue.PushBack(tree.Root)
	level := 0
	nodesInLevel := 1
	
	for queue.Len() > 0 {
		nextLevelCount := 0
		fmt.Printf("Уровень %d: ", level)
		
		for i := 0; i < nodesInLevel; i++ {
			if queue.Len() == 0 {
				break
			}
			
			front := queue.Front()
			current := front.Value.(*TNode)
			queue.Remove(front)
			
			if current != nil {
				fmt.Printf("%s ", current.Data)
				queue.PushBack(current.Left)
				queue.PushBack(current.Right)
				nextLevelCount += 2
			} else {
				fmt.Print("NULL ")
				queue.PushBack(nil)
				queue.PushBack(nil)
				nextLevelCount += 2
			}
		}
		
		fmt.Println()
		level++
		nodesInLevel = nextLevelCount
		
		// Проверяем, есть ли ненулевые узлы в следующем уровне
		hasNonNull := false
		tempQueue := list.New()
		for e := queue.Front(); e != nil; e = e.Next() {
			if e.Value.(*TNode) != nil {
				hasNonNull = true
				break
			}
		}
		tempQueue.Init()
		
		if !hasNonNull {
			break
		}
	}
}

func TPREORDER(tree *Tree) {
	if tree.Root == nil {
		fmt.Println("Дерево пусто!")
		return
	}
	
	fmt.Print("Pre-order: ")
	stack := list.New()
	stack.PushBack(tree.Root)
	
	for stack.Len() > 0 {
		back := stack.Back()
		current := back.Value.(*TNode)
		stack.Remove(back)
		fmt.Printf("%s ", current.Data)
		
		if current.Right != nil {
			stack.PushBack(current.Right)
		}
		if current.Left != nil {
			stack.PushBack(current.Left)
		}
	}
	fmt.Println()
}

func TINORDER(tree *Tree) {
	if tree.Root == nil {
		fmt.Println("Дерево пусто!")
		return
	}
	
	fmt.Print("In-order: ")
	stack := list.New()
	current := tree.Root
	
	for current != nil || stack.Len() > 0 {
		for current != nil {
			stack.PushBack(current)
			current = current.Left
		}
		
		back := stack.Back()
		current = back.Value.(*TNode)
		stack.Remove(back)
		fmt.Printf("%s ", current.Data)
		
		current = current.Right
	}
	fmt.Println()
}

func TPOSTORDER(tree *Tree) {
	if tree.Root == nil {
		fmt.Println("Дерево пусто!")
		return
	}
	
	fmt.Print("Post-order: ")
	stack1 := list.New()
	stack2 := list.New()
	stack1.PushBack(tree.Root)
	
	for stack1.Len() > 0 {
		back := stack1.Back()
		current := back.Value.(*TNode)
		stack1.Remove(back)
		stack2.PushBack(current)
		
		if current.Left != nil {
			stack1.PushBack(current.Left)
		}
		if current.Right != nil {
			stack1.PushBack(current.Right)
		}
	}
	
	for stack2.Len() > 0 {
		back := stack2.Back()
		fmt.Printf("%s ", back.Value.(*TNode).Data)
		stack2.Remove(back)
	}
	fmt.Println()
}

func TCLEAR(tree *Tree) {
	clearTree(tree.Root)
	tree.Root = nil
	tree.Size = 0
	fmt.Println("Дерево очищено")
}
